File: lintcode/merge_two_sorted_lists/py/merge_two_sorted_lists.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/merge-two-sorted-lists
@Language: Python
@Datetime: 17-02-15 21:56
'''

"""
Definition of ListNode
class ListNode(object):
    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param two ListNodes
    @return a ListNode
    """
    def mergeTwoLists(self, l1, l2):
        dummy = ListNode(0)
        dest  = dummy
        
        while l1 != None and l2 != None:
            if l1.val > l2.val:
                dest.next = l2
                l2        = l2.next
            else:
                dest.next = l1
                l1        = l1.next
                
            dest = dest.next
            
        while l1 != None:
            dest.next = l1
            l1        = l1.next
            dest      = dest.next
        
        while l2 != None:
            dest.next = l2
            l2        = l2.next
            dest      = dest.next
        
        return dummy.next

File: lintcode/easy/merge_two_sorted_lists/cpp/merge_two_sorted_lists.cpp
/*
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/merge-two-sorted-lists
@Language: C++
@Datetime: 17-02-15 23:17
*/

/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param ListNode left is the head of the linked list
     * @param ListNode right is the head of the linked list
     * @return: ListNode head of linked list
     */
    ListNode *mergeTwoLists(ListNode *left, ListNode *right)
    {
        ListNode dummy{ 0 }, *tail{ &dummy };
        
        while (left != nullptr && right != nullptr)
        {
            if (left->val > right->val) {
                tail->next = right;
                right      = right->next;
            } else {
                tail->next = left;
                left       = left->next;
            }
            
            tail = tail->next;
        }
        
        while (left != nullptr) {
            tail->next = left;
            left       = left->next;
            tail       = tail->next;
        }
        
        while (right != nullptr) {
            tail->next = right;
            right      = right->next;
            tail       = tail->next;
        }
        
        return dummy.next;
    }
};